ee/tabby-webserver/src/service/repository.rs

```diff
@@ -37,6 +41,62 @@ impl RepositoryService for DbConn {
             .await?;
         Ok(true)
     }
+
+    async fn search_files(
+        &self,
+        name: String,
+        pattern: String,
+        top_n: usize,
+    ) -> Result<Vec<FileEntry>> {
+        if pattern.trim().is_empty() {
+            return Ok(vec![]);
+        }
+        let git_url = self.get_repository_by_name(name.clone()).await?.git_url;
+        let config = RepositoryConfig::new_named(name, git_url);
+        let matching = match_pattern(&config.dir(), &pattern, top_n)
+            .await
+            .map_err(anyhow::Error::from)?;
+        Ok(matching)
+    }
+}
+
+async fn match_pattern(
+    base: &Path,
+    pattern: &str,
+    limit: usize,
+) -> Result<Vec<FileEntry>, anyhow::Error> {
+    let mut nucleo = nucleo::Matcher::new(nucleo::Config::DEFAULT.match_paths());
+    let needle = nucleo::pattern::Pattern::new(
+        pattern,
+        nucleo::pattern::CaseMatching::Ignore,
+        nucleo::pattern::Normalization::Smart,
+        nucleo::pattern::AtomKind::Fuzzy,
+    );
+    let mut scored_entries: Vec<(_, _)> = Walk::new(base)
+        .filter_map(|path| {
+            let entry = path.ok()?;
+            let r#type = if entry.file_type().map(|x| x.is_dir()).unwrap_or_default() {
+                "dir".into()
+            } else {
+                "file".into()
+            };
+            let path = entry
+                .into_path()
+                .strip_prefix(base)
+                .ok()?
+                .to_string_lossy()
+                .into_owned();
+            let haystack: nucleo::Utf32String = path.clone().into();
+            let score = needle.score(haystack.slice(..), &mut nucleo);
+            score.map(|score| (score, FileEntry { r#type, path }))
+        })
+        .take(limit)
```

Comment:
This selects values prior to sorting, which may cause badly-scored results to be the only ones returned if there are many matches.

Reply:
it is mostly a greedy prior - we might adjust it to something like 3xlimit in futute